### ISSUE TITLE
Source-code package management

### ISSUE THREAD

**Issue by nfeske** [genodelabs/genode#1082]
In order to make Genode easier to conquer for new users and developers, we need to rethink the way how 3rd-party code is integrated with Genode.

Currently, there are two parts of our tooling that interfaces with 3rd-party code: The integration with Genode's source tree ("make prepare" mechanism provided by repository-local makefiles), and the building/integration ("make run/..." provided by the build system and run tool). Until now, both mechanisms served us well. But with the number of 3rd-party packages steadily increasing, we are facing the following issues, which might slowly become annoyances:
- There are multiple repositories that deal with downloading and integrating 3rd-party code in different ways, namely libports, ports, ports-foc, base-<kernel>, dde_ipxe, dde_rump, dde_linux, dde_oss, qt4. Even though all makefiles contained in those repositories contain the "prepare" and "clean" rules, they are inconsistent with regard to the handling of corner cases, to updating of packages, and with the use of additional arguments ("PKG=").
- The build system does not reliably detect the cases where 3rd-party code is not yet prepared. Instead, it produces confusing error messages.
- Sometimes the prepared version of the 3rd-party code is outdated and the build expects a different version. This happens particularly when changing branches. For a library that was installed with a certain version, this isn't so much of a problem because the corresponding contrib directory is suffixed with its version. But if the 3rd-party code comes from a git repository, the contrib directory lacks a version suffix. In some cases, the content of the contrib directory is updated quite often (e.g., base-nova). The build system should detect such inconsistencies and respond with a meaningful error message.
- There are no dependencies between 3rd-party packages, which can be quite annoying when starting to explore Genode. For example, for building a Qt application, a lot of further libraries (freetype, libpng...) need to be prepared. Our tools should take care of preparing those dependencies automatically.
- The `<repository>/ports/*.mk` files contain a lot of boiler-plate rules. For example, the rules for downloading files via wget, or the rules for checking signatures.
- The downloaded archives and the extracted 3rd-party code are located within the various repositories (in the `download/` and `contrib/` subdirectories). This makes the use of search tools like grep very inefficient when we want to search only in Genode's source code, excluding 3rd-party sources (with all packages prepared, there are more than 6 GiB of 3rd-party code). For this reason, most regular Genode developers have crafted some special shell aliases for filtered search operations. But this should not be the way to go.

**Comment by ehmry** [comment]
I think I can say I have a Nix port that is complete enough to provide the tools to implement package management, but the more nuanced task of finding the best use of the tools remains. I also really need to write some documentation on the changes I made.

Because the essence of Nix is creating file-system objects, I created a VFS plugin that evaluates Nix language expressions and returns a file system session or rom connection to the result of the evaluation. I have two very simple examples of how this would work, ``run/vfs_nix`` and ``run/vfs_nix_rom``:
https://github.com/ehmry/genode-nix/blob/master/run/vfs_nix_rom.run#L269
https://github.com/ehmry/genode-nix/blob/master/expr/cli_monitor.nix
https://github.com/ehmry/genode-nix/blob/master/run/vfs_nix.run#L172
https://github.com/ehmry/genode-nix/blob/master/expr/noux-result.nix

Using Nix as a string manipulation tool has shown good results in generating configurations, but I have had problems driving Noux with Nix on real hardware within multiple nested inits. That I won't know more about until my serial logging hardware arrives.

Norman and I have discussed offline the prospect of generating configurations for cli_monitor, so that will probably be my focus for experimentation. I would like to start compiling Genode components, but I am only interested in doing this outside of QEMU, so like I said I've got to wait on some hardware.

My repo containing the Nix components:
https://github.com/ehmry/genode-nix

My Genode branch for building against the external repo:
https://github.com/ehmry/genode/tree/nix-port

The implementation details of porting have overwhelmed me and drained my imagination for the moment, so if anyone has suggestions on something I should try and create with Nix, let me know.

**Comment by ehmry** [comment]
There has been a change of plan, the nix VFS plugin hasn't been a good fit, so I'm now working with a server that rewrites ROM and File_system session arguments with the result of a function applied to those arguments. This is probably more useful than the VFS plugin because it is very good at producing configurations. http://nixpaste.lbr.uno/WT3aFCCi?nix

**Comment by nfeske** [comment]
As a side project during the past few weeks, I experimented with different ideas of how to extract "source packages" from the source tree and how to interface them. I pushed the result to a new repository: https://github.com/nfeske/genode-pkg

The thought process while doing the experimentation helped me a lot to get a tangible feeling for the "packaging" problem in the context of Genode. @ehmry  Now that the clouds begin to lift in my head, it might be good time to pick up out past discussion to see how Nix fits into the picture - maybe at the Hack'n'Hike?
